src/file_chooser.rs

```diff
@@ -0,0 +1,474 @@
+        title: &str,
+        options: OpenFileOptions,
+    ) -> PortalResponse<FileChooserResult> {
+        self.run(
+            handle,
+            app_id,
+            parent_window,
+            title,
+            FileChooserOptions::OpenFile(options),
+        )
+        .await
+    }
+
+    async fn save_file(
+        &self,
+        #[zbus(connection)] connection: &zbus::Connection,
+        handle: zvariant::ObjectPath<'_>,
+        app_id: &str,
+        parent_window: &str,
+        title: &str,
+        options: SaveFileOptions,
+    ) -> PortalResponse<FileChooserResult> {
+        self.run(
+            handle,
+            app_id,
+            parent_window,
+            title,
+            FileChooserOptions::SaveFile(options),
+        )
+        .await
+    }
+
+    async fn save_files(
+        &self,
+        #[zbus(connection)] connection: &zbus::Connection,
+        handle: zvariant::ObjectPath<'_>,
+        app_id: &str,
+        parent_window: &str,
+        title: &str,
+        options: SaveFilesOptions,
+    ) -> PortalResponse<FileChooserResult> {
+        self.run(
+            handle,
+            app_id,
+            parent_window,
+            title,
+            FileChooserOptions::SaveFiles(options),
+        )
+        .await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Msg {
+    DialogMessage(DialogMessage),
+    DialogResult(DialogResult),
+}
+
+#[derive(Clone)]
+pub(crate) struct Args {
+    pub handle: zvariant::ObjectPath<'static>,
+    pub app_id: String,
+    pub parent_window: String,
+    pub title: String,
+    pub options: FileChooserOptions,
+    pub tx: Sender<PortalResponse<FileChooserResult>>,
+}
+
+pub(crate) fn view(portal: &CosmicPortal, id: window::Id) -> cosmic::Element<crate::app::Msg> {
+    portal.file_chooser_dialog.as_ref().unwrap().view(id)
+}
+
+pub fn update_msg(
+    portal: &mut CosmicPortal,
+    msg: Msg,
+) -> cosmic::Command<app::Message<crate::app::Msg>> {
+    match msg {
+        Msg::DialogMessage(dialog_msg) => portal
+            .file_chooser_dialog
+            .as_mut()
+            .unwrap()
+            .update(dialog_msg),
+        Msg::DialogResult(dialog_res) => {
+            log::debug!("file chooser result {:?}", dialog_res);
+            let args = portal.file_chooser_args.take().unwrap();
+            let dialog = portal.file_chooser_dialog.take().unwrap();
+            let response = match dialog_res {
+                DialogResult::Cancel => PortalResponse::Cancelled,
+                DialogResult::Open(paths) => {
+                    let mut uris = Vec::with_capacity(paths.len());
+                    for path in paths {
+                        match url::Url::from_file_path(&path) {
+                            Ok(url) => uris.push(url.to_string()),
+                            Err(()) => {
+                                log::error!("failed to convert to URL: {:?}", path);
+                                //TODO: return error?
```

Comment:
If `uris` is empty after this step, maybe `PortalResponse::Other` could be returned?

Reply:
Fixed